sdk/trace/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,258 @@
+const (
+	offsetExponentMask = 0x7ff0000000000000
+	offsetExponentBias = 1023
+	significandBits    = 52
+)
+
+// expFromFloat64 returns floor(log2(x)).
+func expFromFloat64(x float64) int {
+	return int((math.Float64bits(x)&offsetExponentMask)>>significandBits) - offsetExponentBias
+}
+
+// expToFloat64 returns 2^x.
+func expToFloat64(x int) float64 {
+	return math.Float64frombits(uint64(offsetExponentBias+x) << significandBits)
+}
+
+// splitProb returns the two values of log-adjusted-count nearest to p
+// Example:
+//
+//   splitProb(0.375) => (2, 1, 0.5)
+//
+// indicates to sample with probability (2^-2) 50% of the time
+// and (2^-1) 50% of the time.
+func splitProb(p float64) (int, int, float64) {
+	// Take the exponent and drop the significand to locate the
+	// smaller of two powers of two.
+	exp := expFromFloat64(p)
+
+	// Low is the smaller of two log-adjusted counts, the negative
+	// of the exponent computed above.
+	low := -exp
+	// High is the greater of two log-adjusted counts (i.e., one
+	// less than low, a smaller adjusted count means a larger
+	// probability).
+	high := low - 1
+
+	// Return these to probability values and use linear
+	// interpolation to compute the required probability of
+	// choosing the low-probability Sampler.
+	lowP := expToFloat64(-low)
+	highP := expToFloat64(-high)
+	lowProb := (highP - p) / (highP - lowP)
+
+	return low, high, lowProb
+}
+
+// tidSamplerForLogAdjustedCount
+func tidSamplerForLogAdjustedCount(logAdjustedCount int) trace.Sampler {
+	return trace.TraceIDRatioBased(expToFloat64(-logAdjustedCount))
+}
+
+func (s *Sampler) ShouldSample(params trace.SamplingParameters) trace.SamplingResult {
+	state := s.current.Load().(*window)
+	now := s.nowfunc()
+
+	if now.Sub(state.start) >= s.interval {
+		// If the window has expired, update it and re-load.
+		state.compute.Do(func() {
+			s.updateWindow(state, now)
+		})
+		state = s.current.Load().(*window)
+	}
+
+	// Count the span in this window's rate estimate.
+	_ = atomic.AddInt64(&state.count, 1)
+
+	// Compare a uniform random with lowProb, choose either the
+	// low or high probability TraceIDRatio Sampler.
+	var tid trace.Sampler
+	if rand.Float64() < state.lowProb {
+		tid = state.low
+	} else {
+		tid = state.high
+	}
+
+	return tid.ShouldSample(params)
+}
+
+func (s *Sampler) Description() string {
+	return fmt.Sprintf("RateLimited{%g}", s.targetRate)
+}
+
+func (s *Sampler) updateWindow(expired *window, now time.Time) {
+	// Capture the actual count and the corresponding interval
+	// that was measured since the probability was last updated.
+	count := atomic.LoadInt64(&expired.count)
+	duration := now.Sub(expired.start)
+
+	// Combine the new data and the old data.  In Bayesian terms,
+	// this is justified by modelling the arrival of spans as a
+	// Poisson process.  The maximum-a-posteriori estimate of the
+	// rate based on the observed data equals totalCount divided
+	// by totalDuration.
+	totalCount := count + s.priorCount
```

Comment:
The problem I see here is that the sampling rate will get less and less adaptive over time, as the current count and duration values have less and less influence on the the totalCount and the totalDuration and therefore also on the predicted rate. For example, if there is a continuous load of 100 spans per second over weeks and suddenly drops to 1 span per second, it will take very long until the sampling rate is adjusted correspondingly. I think there is the need of some forgetting mechanism. I think a second parameter is needed (in addition to the target rate) that controls the forgetting factor / adaption rate.

Reply:
I think I've misled you with a poorly chosen variable name. The `totalCount` and `totalDuration` used for estimation should be the sum over two windows every time (assuming the initial state has a prior of 0 count, 0 duration). The `s.priorCount` and `s.priorDuration` are set to this window's count and duration for use in the next round. 

The predicted rate is just the average rate over the two windows. I suppose this is just a heuristic -- we could use exponential averaging or any number of techniques to forecast the upcoming rate based on recently observed rates.